### PR TITLE
TEL review of day 4 homework

### PR DESCRIPTION
**Pull request by tomlatham**
Good attempt at a tricky problem!  The main issue was in the code that inserted X's in between identical characters in digraphs did not work as intended - when comparing characters you were comparing the first character in a digraph with the second character of the previous digraph.
The application of the rules of the cipher were good for same-row and same-column cases but for the rectangle case you had a small bug - see the line-by-line comments.
Writing some tests would probably have helped to iron out some of these issues.  I've added a single test example but really one ought to have a set of tests that cover a range of edge cases.

### CODE REVIEW

File: MPAGSCipher/PlayfairCipher.cpp [hunk 1]
```diff
@@ -12,155 +12,118 @@
...
-	    }
-	}
-  
-      // If size of input is odd, add a trailing Z
-      if (j!=0)
-	{
-	  text = text + 'Z';
-	}  
+  //Encrypt or decrypt - set the shift accordingly
+  const std::size_t shift { (cipherMode == CipherMode::Encrypt) ? 1u : 4u };
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: This line avoids the need to duplicate vast swathes of code.
Depending on the encrypt/decrypt mode we set the shift within the row/column to be either +1 or -1 (or more technically, -1 + 5 = 4).

File: MPAGSCipher/PlayfairCipher.cpp [hunk 2]
```diff
@@ -12,155 +12,118 @@
...
 
-      const size_t sizeText = text.size();
+  // Change J -> I
+  auto changeJI = [] (char elem)
+    {
+      if (elem =='J') {
+        return 'I';
+      } else {
+        return elem;
+      }
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Best to always include braces - search for the Apple SSL "goto fail" bug for an example of where not doing so wreaked havoc!

File: MPAGSCipher/PlayfairCipher.cpp [hunk 3]
```diff
@@ -12,155 +12,118 @@
...
-      // If repeated chars in a digraph add an X or  Q if X
-
-      const size_t size = inputText.size();
-      
-      int j =0;
-
-      for(size_t i=1; i < size; i++)
-	{
-	  j = i%2;
-	  if (j==0)
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: The combination of these conditions is equivalent to starting from 0 and looping with a step of 2.  I wonder, looking at this again to write these comments, did you mean to check if (j==1) here?

File: MPAGSCipher/PlayfairCipher.cpp [hunk 4]
```diff
@@ -12,155 +12,118 @@
...
-	    {
-	      if(inputText[i] == inputText[i-1])
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: So here you're checking against the previous character, which is not in the same digraph as this one.  If you had checked for j == 1 then this might would have been the right comparison.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 5]
```diff
@@ -12,155 +12,118 @@
...
-		{
-		  if(inputText[i-1] != 'X')
-		    {
-		      text[i] = 'X';
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: This acts to replace the character with X, not to insert an X before it.  So "LL" -> "XL", whereas it should have become "LXL".

File: MPAGSCipher/PlayfairCipher.cpp [hunk 6]
```diff
@@ -12,155 +12,118 @@
...
-	    }
-	  else if (coordinate1.second == coordinate2.second)
-	    {
-	      coordinate1.first = (coordinate1.first+1)%5;
-	      coordinate2.first = (coordinate2.first+1)%5;
-	    }
-	  else
-	    {
-	      coordinate1.second = coordinate2.second;
-	      coordinate2.second = coordinate1.second;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Here you need to use a temporary variable to store coodinate1.second, since you overwrite it in the first line, then try to use it in the second.  Something like the following would have worked:
```C++
auto tmp = coordinate1.second;
coordinate1.second = coordinate2.second;
coordinate2.second = tmp;
```
Or, as I've done below, you can just use `std::swap`.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 7]
```diff
@@ -12,155 +12,118 @@
...
+          coordinate2.second = (coordinate2.second+shift)%5;
+        }
+      else if (coordinate1.second == coordinate2.second)
+        {
+          coordinate1.first = (coordinate1.first+shift)%5;
+          coordinate2.first = (coordinate2.first+shift)%5;
+        }
+      else
+        {
+          std::swap(coordinate1.first, coordinate2.first);
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: NB, since you've stored your coordinates as (column,row), you need to swap the first element rather than the second.

File: MPAGSCipher/PlayfairCipher.cpp [hunk 8]
```diff
@@ -198,23 +161,13 @@ void PlayFairCipher::setKey(const std::string& key)
...
+  for (std::size_t i = 0; i < 25; i++)
     {
-      std::pair<int, int> coord{i%5,i/5};
-      auto positionItem{ std::make_pair(coord, key_[i])};
-      auto letterItem{ std::make_pair(key_[i], coord)};
-      positionmap_.insert(positionItem);
-      lettermap_.insert(letterItem);
+      PlayfairCoords coord{i%5,i/5};
+      positionmap_.insert( std::make_pair(coord, key_[i]) );
+      lettermap_.insert( std::make_pair(key_[i], coord) );
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.cpp]: Simplify this a bit by using the type definition (see the changes to the header file) for the coordinate construction.

File: MPAGSCipher/PlayfairCipher.hpp [hunk 9]
```diff
@@ -42,28 +42,22 @@ class PlayFairCipher {
...
-   *
-   *\param key_ the sze_t parameter used to store the cipher key
-   */
+  /// The cipher key phrase
   std::string key_{'A'};
   
-  using Coord2StringMap = std::map<std::pair<std::size_t, std::size_t>, char>;
-  using String2CoordMap = std::map<char, std::pair<std::size_t, std::size_t>>;
+  /// Type definition for the coordinates in the 5x5 table
+  using PlayfairCoords = std::pair<std::size_t, std::size_t>;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: This is the most useful type definition to make since this is used quite frequently in the implementation code.
Declaring it here in the private part of the class means that we can use it within the implementation of the class functions but not outside the class, which is appropriate since it is an implementation detail and doesn't appear in the interface.

File: MPAGSCipher/PlayfairCipher.hpp [hunk 10]
```diff
@@ -42,28 +42,22 @@ class PlayFairCipher {
...
+
+  /// Type definition for the map relating the coordinates to the key letter
+  using Coord2StringMap = std::map<PlayfairCoords, char>;
 
+  /// Type definition for the map relating the key letter to the coordinates
+  using String2CoordMap = std::map<char, PlayfairCoords>;
```
Comment (tomlatham) [inline MPAGSCipher/PlayfairCipher.hpp]: Making these type definitions isn't actually so useful since they are used only once, directly below these lines.
